toolchain/check/handle_impl.cpp

```diff
@@ -206,22 +259,48 @@ static auto BuildImplDecl(Context& context, Parse::AnyImplDeclId node_id,
 
   // TODO: Check for an orphan `impl`.
 
-  // TODO: Check parameters. Store them on the `Impl` in some form.
-  static_cast<void>(params_id);
+  // Add the impl declaration.
+  SemIR::ImplDecl impl_decl = {.impl_id = SemIR::ImplId::Invalid,
+                               .decl_block_id = decl_block_id};
+  auto impl_decl_id =
+      context.AddPlaceholderInst(SemIR::LocIdAndInst(node_id, impl_decl));
+
+  SemIR::Impl impl_info = {
+      name_context.MakeEntityWithParamsBase(name, impl_decl_id,
+                                            /* is_extern=*/false,
+                                            SemIR::LibraryNameId::Invalid),
+      {.self_id = self_type_id, .constraint_id = constraint_type_id}};
 
   // Add the impl declaration.
   // TODO: Does lookup in an impl file need to look for a prior impl declaration
   // in the api file?
-  auto impl_id = context.impls().LookupOrAdd(self_type_id, constraint_type_id);
-  SemIR::ImplDecl impl_decl = {.impl_id = impl_id,
-                               .decl_block_id = decl_block_id};
-  auto impl_decl_id = context.AddInst(node_id, impl_decl);
+  auto& impl_lookup =
+      context.impls().LookupBucket(impl_info.self_id, impl_info.constraint_id);
```

Comment:
```suggestion
  auto& impl_bucket =
      context.impls().LookupBucket(impl_info.self_id, impl_info.constraint_id);
```

For me, "lookup" is a verb, "bucket" is the noun for the returned object.

Reply:
The reference could also be invalidated by the call to `FinishGenericDecl` triggering an import. Changed to redo the lookup on insert to avoid that.

Reply:
Renamed to `lookup_bucket`.